packages/cli/src/auth/index.ts

```diff
@@ -1,86 +1,243 @@
+
+  // Get or generate a localwallet.
+  const wallet = await getOrGenerateLocalWallet(credsConfigPath, cliWalletPath);
+  const walletAddress = await wallet.getAddress();
+  const auth = new ThirdwebAuth(wallet, "localhost:3000");
+
+  // Generate the login payload to pass to the dashboard.
+  const loggedIn = await auth.login({
+    domain: "localhost:3000",
+    address: walletAddress,
+  });
+
+  // In this case the state is the loggedIn object.
+  const ourState = encodeURIComponent(JSON.stringify(loggedIn));
+  const urlToOpen =
+    // `https://thirdweb.com/cli/login?from=cli&#${ourState}`;
+    // `https://thirdweb-www-git-mariano-api-keys-sign-in.thirdweb-preview.com/cli/login?from=cli&#${ourState}`;
+    `http://localhost:3000/cli/login?from=cli&#${ourState}`;
+
+  let server: http.Server;
+  let loginTimeoutHandle: NodeJS.Timeout;
+  const timerPromise = new Promise<void>((resolve, reject) => {
+    loginTimeoutHandle = setTimeout(() => {
+      logger.error("Timed out waiting, please try again.");
+      server.close();
+      clearTimeout(loginTimeoutHandle);
+      reject(new Error("Timed out waiting"));
+    }, 120000);
+  });
+
+  const loginPromise = new Promise<string>((resolve, reject) => {
+    server = http.createServer(async (req, res) => {
+      function finish(error?: Error) {
+        clearTimeout(loginTimeoutHandle);
+        server.close((closeErr?: Error) => {
+          if (error || closeErr) {
+            reject(error || closeErr);
+          }
+        });
+      }
+      res.setHeader("Access-Control-Allow-Origin", "http://localhost:3000");
+      res.setHeader("Access-Control-Allow-Methods", "GET");
+
+      assert(req.url, "This request doesn't have a URL");
+      const { pathname, query } = url.parse(req.url, true);
+      switch (pathname) {
+        case "/auth/callback": {
+          if (query.token) {
+            const token = Array.isArray(query.token) ? query.token[0] : query.token;
+            const theirState = Array.isArray(query.state)
+              ? query.state[0]
+              : query.state;
+
+            // Check if the payload sent back is the same as the one we sent.
+            if (theirState !== JSON.stringify(loggedIn)) {
+              res.writeHead(400, { "Content-Type": "text/plain" }); // send a 400 response
+              res.end("Unauthorized request, state mismatch", () => {
+                finish(new Error("Unauthorized request, state mismatch"));
+              });
+              reject(
+                new Error(chalk.red("Unauthorized request, state mismatch")),
+              );
+            } else {
+              // Save the token to the config file.
+              const configJson = JSON.parse(fs.readFileSync(credsConfigPath, "utf-8"));
+              fs.writeFileSync(credsConfigPath, JSON.stringify({
+                ...configJson,
+                token
+              }), {
+                encoding: "utf8",
+                mode: 0o600,
```

Comment:
what's this mode do and why do we need it?

Reply:
This makes sure that the permissions of the file being written is set to 600, to ensure only admins of that computer will be able to read/write from it.